File: packages/expo-task-manager/android/src/main/java/expo/modules/taskManager/TaskManagerModule.java
package expo.modules.taskManager;

import android.content.Context;
import android.os.Handler;

import expo.modules.core.ExportedModule;
import expo.modules.core.ModuleRegistry;
import expo.modules.core.Promise;
import expo.modules.core.interfaces.ExpoMethod;
import expo.modules.interfaces.taskManager.TaskManagerInterface;
import expo.modules.interfaces.taskManager.TaskServiceInterface;

import java.util.HashMap;
import java.util.Map;

public class TaskManagerModule extends ExportedModule {

  private TaskServiceInterface mTaskService;
  private TaskManagerInterface mTaskManagerInternal;

  public TaskManagerModule(Context context) {
    super(context);
  }

  @Override
  public String getName() {
    return "ExpoTaskManager";
  }

  @Override
  public Map<String, Object> getConstants() {
    Map<String, Object> constants = new HashMap<>();
    constants.put("EVENT_NAME", TaskManagerInterface.EVENT_NAME);
    return constants;
  }

  @Override
  public void onCreate(ModuleRegistry moduleRegistry) {
    mTaskService = moduleRegistry.getSingletonModule("TaskService", TaskServiceInterface.class);
    mTaskManagerInternal = moduleRegistry.getModule(TaskManagerInterface.class);
  }

  //region Expo methods
  
  @ExpoMethod
  public void isAvailableAsync(final Promise promise) {
    promise.resolve(mTaskService != null);
  }

  @ExpoMethod
  public void notifyTaskFinishedAsync(String taskName, Map<String, Object> response, final Promise promise) {
    if (!checkTaskService(promise)) {
      return;
    }
    mTaskService.notifyTaskFinished(taskName, getAppScopeKey(), response);
    promise.resolve(null);
  }

  @ExpoMethod
  public void isTaskRegisteredAsync(String taskName, final Promise promise) {
    if (!checkTaskService(promise)) {
      return;
    }
    promise.resolve(mTaskService.hasRegisteredTask(taskName, getAppScopeKey()));
  }

  @ExpoMethod
  public void getTaskOptionsAsync(String taskName, final Promise promise) {
    if (!checkTaskService(promise)) {
      return;
    }
    promise.resolve(mTaskService.getTaskOptions(taskName, getAppScopeKey()));
  }

  @ExpoMethod
  public void getRegisteredTasksAsync(final Promise promise) {
    if (!checkTaskService(promise)) {
      return;
    }
    promise.resolve(mTaskService.getTasksForAppScopeKey(getAppScopeKey()));
  }

  @ExpoMethod
  public void unregisterTaskAsync(String taskName, final Promise promise) {
    if (!checkTaskService(promise)) {
      return;
    }
    try {
      mTaskService.unregisterTask(taskName, getAppScopeKey(), null);
      promise.resolve(null);
    } catch (Exception e) {
      promise.reject(e);
    }
  }

  @ExpoMethod
  public void unregisterAllTasksAsync(final Promise promise) {
    if (!checkTaskService(promise)) {
      return;
    }
    try {
      mTaskService.unregisterAllTasksForAppScopeKey(getAppScopeKey());
      promise.resolve(null);
    } catch (Exception e) {
      promise.reject(e);
    }
  }

  @ExpoMethod
  public void startObserving(final Promise promise) {
    Handler handler = new Handler();
    handler.postDelayed(new Runnable() {
      @Override
      public void run() {
        if (mTaskManagerInternal != null) {
          mTaskManagerInternal.flushQueuedEvents();
        }
      }
    }, 1000);
    promise.resolve(null);
  }

  @ExpoMethod
  public void stopObserving(final Promise promise) {
    // nothing
    promise.resolve(null);
  }

  //endregion
  //region helpers

  private String getAppScopeKey() {
    return mTaskManagerInternal.getAppScopeKey();
  }

  private boolean checkTaskService(final Promise promise) {
    if (mTaskService == null) {
      promise.reject(TaskManagerInterface.ERR_TASK_SERVICE_NOT_FOUND, "Unable to find TaskService singleton module in module registry.");
      return false;
    }
    return true;
  }

  //endregion
}

File: packages/expo-task-manager/android/src/main/java/expo/modules/taskManager/TaskManagerModule.kt
package expo.modules.taskManager

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import expo.modules.core.errors.ModuleNotFoundException
import expo.modules.interfaces.taskManager.TaskManagerInterface
import expo.modules.interfaces.taskManager.TaskServiceInterface
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.lang.ref.WeakReference

class TaskManagerModule : Module() {
  private val _taskService: TaskServiceInterface? by lazy {
    appContext.legacyModuleRegistry.getSingletonModule("TaskService", TaskServiceInterface::class.java)
  }
  private val taskService: TaskServiceInterface
    get() = _taskService
      ?: throw ModuleNotFoundException(TaskServiceInterface::class.java.toString())

  private val taskManagerInternal: TaskManagerInterface? by lazy {
    appContext.legacyModule()
  }

  override fun definition() = ModuleDefinition {
    Name("ExpoTaskManager")

    Events(TaskManagerInterface.EVENT_NAME)

    Constants(
      "EVENT_NAME" to TaskManagerInterface.EVENT_NAME
    )

    OnCreate {
      // Slightly hacky way to be able to emit events using the new event emitter from legacy modules.
      val weakModule = WeakReference(this@TaskManagerModule)
      val emitEvent = { name: String, body: Bundle ->
        try {
          // It may throw, because RN event emitter may not be available
          // we can just ignore those cases
          weakModule.get()?.sendEvent(name, body)
        } catch (error: Throwable) {
          Log.e("ExpoTaskManager", "Failed to emit event $name using the module's event emitter: ${error.message}")
        }
        Unit
      }

      // For compatibility reasons, we don't want to edit TaskManagerInterface, as it's in the expo-modules-core package.
      // There is only one usage of the TaskManagerInterface and it's the TaskManagerInternalModule, so we can safely cast
      // to it and set it's emitEventWrapper.
      (appContext.legacyModule<TaskManagerInterface>() as? TaskManagerInternalModule)?.setEmitEventWrapper(emitEvent)
    }

    AsyncFunction<Boolean>("isAvailableAsync") {
      return@AsyncFunction true
    }

    AsyncFunction("notifyTaskFinishedAsync") { taskName: String, response: Map<String, Any?> ->
      taskService.notifyTaskFinished(taskName, appScopeKey, response)
    }

    AsyncFunction("isTaskRegisteredAsync") { taskName: String ->
      taskService.hasRegisteredTask(taskName, appScopeKey)
    }

    AsyncFunction("getTaskOptionsAsync") { taskName: String ->
      taskService.getTaskOptions(taskName, appScopeKey)
    }

    AsyncFunction<List<Bundle>>("getRegisteredTasksAsync") {
      taskService.getTasksForAppScopeKey(appScopeKey)
    }

    AsyncFunction("unregisterTaskAsync") { taskName: String ->
      taskService.unregisterTask(taskName, appScopeKey, null)
    }

    AsyncFunction<Unit>("unregisterAllTasksAsync") {
      taskService.unregisterAllTasksForAppScopeKey(appScopeKey)
    }

    OnStartObserving {
      val handler = Handler(Looper.getMainLooper())
      handler.postDelayed(
        {
          taskManagerInternal?.flushQueuedEvents()
        },
        1000
      )
    }
  }

  private val appScopeKey: String
    get() = (
      taskManagerInternal
        ?: throw ModuleNotFoundException(TaskManagerInterface::class.java.toString())
      ).appScopeKey
}
